contributors/design-proposals/bulk_watch.md

```diff
@@ -0,0 +1,450 @@
+meet to satisfy our needs
+- a single bulk requests has to support multiple resource types (e.g. get a
+node and all pods associated with it)
+- if the bulk requests filters asks only for object of a single resource type,
+the returned result ideally should be compatible with what is returned from
+the regular API
+- the API has to support aggregation if different resource types are served
+by different apiservers
+- clients has to be able to alter their watch subscribtions incrementally (it
+may not be implemented in the initial version though, but has to be designed)
+
+
+# Detailed design
+
+ As stated in above requirements, we need to make bulk operations work across
+different resource types (e.g. watch pod P and secret S within a single watch
+call). To clarify, the initial implementation doesn't need to support it, but
+it has to be possible to extend the implementation without changing the API.
+Spanning multiple resources, resources types or conditions will be more and more
+important for large number of watches. As an example, federation will be adding
+watches for every type it federates. With that in mind, bypassing aggregation
+at the resource type level and going to aggregation over objects with different
+resource types will allow us to more aggresively optimize in the future.
+
+ Moreover, out current REST API doesn't even offer an easy way to handle
+"multiple watches of a given type" within a single request. As a result, instead
+of inventing new top level pattern per type, we can introduce a new resource
+type that follows normal RESTful rules and solves even more generic problem
+of spanning multiple different resource types.
+
+ We will start with introducing a new dedicated API group:
+ ```
+ /apis/bulk.k8s.io/
+ ```
+ that underneath will have a completely separate implementation.
+
+ In this design, we will focus only on bulk list (get) and watch operations.
+Later, we would like to introduce new resources to support bulk create, update
+and delete operations, but that's not part of this design.
+
+ We will start with introducing `list` resource and supporting the following
+two operation:
+```
+POST /apis/bulk.k8s.io/v1/list <body defines filtering>
+POST /apis/bulk.k8s.io/v1/list?watch=1 <body defines filtering>
+```
+ We can't simply make this an http GET request, due to limitations of GET for
+the size (length) of the url (in which we would have to pass filter options).
+This makes the API look similar to our regular API.
+ Note that once we have list fully implemented, adding this kind of watch will
+be faily simple.
+
+ The main drawback of this approach is that it won't allow for dynamic altering
+of watch subscriptions (which we would like to also support).
+As a result, we need another API for watch that will also support incremental
+subscribtions. That's why we will also introduce the following operation:
+```
+websocket /apis/bulk.k8s.io/v1/list?watch=1
```

Comment:
OK - now I see that it's definitely not clear from the doc. I just made this more clear.